pkg/querier/querier.go

```diff
@@ -505,31 +516,68 @@ func (s *sliceSeriesSet) Warnings() storage.Warnings {
 	return nil
 }
 
-func validateQueryTimeRange(ctx context.Context, userID string, startMs, endMs int64, limits *validation.Overrides, maxQueryIntoFuture time.Duration, logger log.Logger) (int64, int64, error) {
-	now := model.Now()
-	startTime := model.Time(startMs)
-	endTime := model.Time(endMs)
-
-	endTime = clampTime(ctx, endTime, maxQueryIntoFuture, now.Add(maxQueryIntoFuture), false, "end", "max query into future", logger)
+func validateQueryTimeRange(userID string, startMs, endMs int64, limits *validation.Overrides, maxQueryIntoFuture time.Duration, logger log.Logger) (int64, int64, error) {
+	now := time.Now().UnixMilli()
+	endMs = clampTime(logger, endMs, now, maxQueryIntoFuture, "max query into future", true)
 
 	maxQueryLookback := limits.MaxQueryLookback(userID)
-	startTime = clampTime(ctx, startTime, maxQueryLookback, now.Add(-maxQueryLookback), true, "start", "max query lookback", logger)
+	startMs = clampTime(logger, startMs, now, -maxQueryLookback, "max query lookback", false)
 
-	if endTime.Before(startTime) {
+	if endMs < startMs {
 		return 0, 0, errEmptyTimeRange
 	}
 
-	return int64(startTime), int64(endTime), nil
+	return startMs, endMs, nil
 }
 
-// Ensure a time is within bounds, and log in traces to ease debugging.
-func clampTime(ctx context.Context, t model.Time, limit time.Duration, clamp model.Time, before bool, kind, name string, logger log.Logger) model.Time {
-	if limit > 0 && ((before && t.Before(clamp)) || (!before && t.After(clamp))) {
-		level.Debug(spanlogger.FromContext(ctx, logger)).Log(
-			"msg", "the "+kind+" time of the query has been manipulated because of the '"+name+"' setting",
-			"original", util.FormatTimeModel(t),
-			"updated", util.FormatTimeModel(clamp))
-		t = clamp
+// clampTime allows for time-limiting query minT and maxT based on configured limits.
```

Comment:
Would be good to add an explanation of what `clampMax` does. I'm also wondering if a name like `isMinT` / `isMaxT` would be a clearer description of what this parameter is.

Reply:
Going even further, this boolean is almost completely changing the behavior of this method. Would it be better to split this method into  `clampMax()` and `clampMin()` methods that each enforce/adjust one thing? The code duplication between those seems like it would be minimal and it might remove some complexity around passing negative durations in.

Reply:
:+1: I think `clampMax` and `clampMin` will be much nicer

Reply:
... though unfortunately I don't believe we can avoid requiring either passing in a negative duration for limit delta, or having the caller indicate a direction to apply the delta in.

I attempted to assume directionality in a `clampMaxTime` implementation like so:
```go
	if refT > maxT {
		// maxT is in the past w.r.t refT; assume limit delta is subtracting from refT
		// this is the query-store-after case
		allowedT = refT - limitDelta.Milliseconds()
	} else {
		// maxT is in the future w.r.t refT; assume limit delta is adding to refT
		// this is the max-query-into-future case
		allowedT = refT + limitDelta.Milliseconds()
	}
	clampedT := math.Min(maxT, allowedT)
```

But when attempting to apply `max-query-into-future` where the front end generated the `maxT` using a `now()` function, the `refT` will still end up greater than the `maxT` by whatever request latency occurred between request generation and when the querier code generated its own `now` to perform the check against - so we never hit the `else` case which was supposed to capture usage of `max-query-into-future` 

I think passing in a negative duration is a clear enough interface for now, particularly with the now-simplified approach and split between `clampMaxTime` and `clampMinTime`.